test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java

```diff
@@ -0,0 +1,156 @@
+import org.elasticsearch.transport.nio.channel.NioSocketChannel;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.concurrent.ThreadFactory;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * The NioGroup is a group of selectors for interfacing with java nio. When it is started it will create the
+ * configured number of socket and acceptor selectors. Each selector will be running in a dedicated thread.
+ * Server connections can be bound using the {@link #bindServerChannel(InetSocketAddress, ChannelFactory)}
+ * method. Client connections can be opened using the {@link #openChannel(InetSocketAddress, ChannelFactory)}
+ * method.
+ *
+ * The logic specific to a particular channel is provided by the {@link ChannelFactory} passed to the method
+ * when the channel is created. This is what allows an NioGroup to support different channel types.
+ */
+public class NioGroup implements AutoCloseable {
+
+    private static final int CREATED = 0;
+    private static final int RUNNING = 1;
+    private static final int STOPPED = 2;
+
+    private final Logger logger;
+
+    private final int acceptorCount;
+    private final ThreadFactory acceptorThreadFactory;
+    private final ArrayList<AcceptingSelector> acceptors;
+    private final BiFunction<Logger, Supplier<SocketSelector>, AcceptorEventHandler> acceptorEventHandlerFunction;
+    private RoundRobinSupplier<AcceptingSelector> acceptorSupplier;
+
+    private final int socketSelectorCount;
+    private final ThreadFactory socketSelectorThreadFactory;
+    private final ArrayList<SocketSelector> socketSelectors;
+    private final Function<Logger, SocketEventHandler> socketEventHandlerFunction;
+    private RoundRobinSupplier<SocketSelector> socketSelectorSupplier;
+
+    private volatile int state = CREATED;
+
+    public NioGroup(Logger logger, ThreadFactory acceptorThreadFactory, int acceptorCount,
+                    BiFunction<Logger, Supplier<SocketSelector>, AcceptorEventHandler> acceptorEventHandlerFunction,
+                    ThreadFactory socketSelectorThreadFactory, int socketSelectorCount,
+                    Function<Logger, SocketEventHandler> socketEventHandlerFunction) {
+        this.logger = logger;
+        this.acceptorThreadFactory = acceptorThreadFactory;
+        this.acceptorEventHandlerFunction = acceptorEventHandlerFunction;
+        this.socketSelectorThreadFactory = socketSelectorThreadFactory;
+        this.acceptorCount = acceptorCount;
+        this.socketSelectorCount = socketSelectorCount;
+        this.socketEventHandlerFunction = socketEventHandlerFunction;
+        acceptors = new ArrayList<>(this.acceptorCount);
+        socketSelectors = new ArrayList<>(this.socketSelectorCount);
+    }
+
+    public synchronized void start() throws IOException {
```

Comment:
if we can prevent this kind of state handling we should. My experience with stuff like this is consistently negative so lets try to move it to the ctor and make it clear once the group is created it's ready to use. I still think we should have an `ensureOpen()` method that checks a boolean if we are still open.